### PR TITLE
[feat](datasink) Add a generic distributed Python DataSink contract

### LINKED ISSUE #646: [feat](datasink) Add a generic distributed Python DataSink contract
### Problem

Vane has native distributed output paths for COPY and statically linked extensions, but it has no public, provider-neutral Python contract for writing relation batches to external services. A connector author must currently perform side effects inside a generic batch callback or add provider-specific runner/native code. Neither route defines binding, retry safety, terminal outcomes, bounded worker results, cleanup, or secret handling.

This is especially unsafe under Ray/FTE: task attempts may be replayed or canceled after a remote side effect, and a losing attempt may still have written data. A generic callback therefore cannot honestly promise exactly-once behavior.

#542 explored a DataSink contract and Turbopuffer adapter in one change, but that implementation predates the current result/finalization and extension-write contracts and couples the reusable execution surface to one provider SDK.

### Proposed outcome

Add a provider-neutral distributed Python DataSink contract:

- Introduce public `DataSink`, `BoundDataSink`, `DataSinkWriter`, `WriteContext`, `WriteResult`, and `WriteSummary` APIs with explicit immutable serialization boundaries.
- Expose a terminal `write_datasink(...)` relation operation that runs through both `local-fast` and Ray.
- Integrate Python DataSink as a new terminal result kind alongside the existing COPY and extension-write paths. Existing COPY and extension behavior must remain unchanged.
- Define explicit commit-protocol and retry capabilities. The initial Ray implementation should accept only immediate sinks that declare replay-safe/idempotent writes; reject unsupported or non-idempotent sinks before side effects begin.
- Make partial visibility explicit: an immediate sink is not an all-or-nothing transaction, and a failed job may already have applied some batches.
- Return only bounded per-task metadata, counts, warnings, and outcome information to the driver. Distinguish applied, aborted, and unknown outcomes without treating an acknowledgement as exactly-once proof.
- Keep credentials out of cloudpickled bind state and plans. Sinks should carry opaque environment/secret references rather than credential values.
- Provide provider-neutral batching/resource limits and a keyed-upsert helper for connectors that require deterministic explicit IDs, including global duplicate-key validation before distributed writes begin.
- Add fake-sink tests for local and Ray execution, zero-row input, task replay/speculation, losing attempts with side effects, partial failure, cleanup, bounded results, serialization, secret redaction, and COPY/extension-write regressions.
- Do not add any provider SDK or concrete Turbopuffer, Qdrant, Milvus, CSV, or file-format adapter in this issue.

### Alternatives and tradeoffs

- Merge #542 as one generic-plus-provider change. This provides a working example sooner, but makes the core contract harder to review and carries an older terminal-finalization design onto current main.
- Use `map_batches` for external writes. This avoids native integration but has no terminal write outcome, retry declaration, bounded result contract, or safe failure boundary.
- Implement a general two-phase transaction protocol immediately. That would support more sinks, but requires durable operation identity, attempt fencing, selected-attempt reconciliation, and provider-specific recovery semantics that the first vector-database adapters do not need.
- Reuse the static-extension write contract from #544. That contract is appropriate for native DuckDB extensions, not Python SDK-backed external services.

### Additional context

- Build this against current `main` and its existing COPY and extension-write finalization contracts from #544; do not replace them with #542's older `FinalizableSinkNode` union.
- #596 concerns Python Relation DML entry-point routing and is related but does not provide this external-service sink contract.
- #543 and #639 remain on the native COPY path and should not depend on Python DataSink.
- #542 is useful prior art for the lifecycle and tests, but the implementation should be separated from Turbopuffer and rebased architecturally rather than cherry-picked unchanged.
Provider follow-ups:

- [ ] #647
- [x] #648
- [x] #649

### PR DESCRIPTION
**Pull request by kaka11chen**
## Summary

- Introduce the provider-neutral `DataSink`, `BoundDataSink`, `DataSinkWorker`, `WriteContext`, `WriteResult`, and `WriteSummary` APIs, exposed through `Relation.write_datasink()`.
- Execute sinks exclusively as long-lived actors: Ray uses `ray_actor`, local/local-fast use `subprocess_actor`, and `worker_count` controls the number of independent sink workers. Each actor lazily opens one worker, reuses it for its batches, aborts it after a failed write, and closes it during graceful query teardown.
- Add a dedicated logical/physical `DATA_SINK` path and distributed finish/result protocol with explicit `APPLIED`, `ABORTED`, and `UNKNOWN` outcomes and bounded protocol payloads.
- Reuse the distributed UDF execution layer for bounded task admission, actor-slot backpressure, and transparent RefBundle input. RefBundles remain an internal transport detail; sink implementations receive a worker-local `pyarrow.Table`.
- Bind sinks with complete native DuckDB Arrow schema conversion, including DECIMAL, UUID, TIMETZ, and ENUM.
- Gracefully close quiesced local-fast sink actors after both successful and failed queries, preserve `DataSinkWorker.close()` failures and graceful-close timeouts as bounded cleanup warnings, and escalate timed-out subprocesses to forced termination.
- Retry incomplete Ray DataSink teardown asynchronously until ownership is released, with per-lifecycle deduplication and identity fencing so an old retry cannot tear down a reused plan ID.
- Stream selected DataSink results through a capacity-one coordinator channel, release each Ray result handle after publication, reject retry losers, and enforce exact count/wire-size limits before retaining each result.
- Add global keyed-upsert validation, environment-secret references, and local, real-Ray, replay, cleanup, serialization, boundary, backpressure, and native distributed coverage.
- Deliberately remove the task-mode API and its compatibility/fallback paths. Provider adapters such as Turbopuffer, Milvus, and Qdrant remain outside this issue.

## Related issue

close: #646

## Documentation impact

- [x] No user-facing documentation update is required.
- [ ] Documentation is updated in this PR or in a linked website PR.
- [ ]  A follow-up issue is required in
      `AstroVela/vane-website`.

## Validation

- Two consecutive static code/semantic review passes completed without findings after the review fixes.
- `git diff --check`
- `scripts/format root --changed --check` — Ruff and clang-format passed
- `CMAKE_BUILD_PARALLEL_LEVEL=32 SKBUILD_BUILD_DIR="$PWD/build/python-release" SKBUILD_CMAKE_BUILD_TYPE=Release uv pip install . --no-build-isolation` — final rebased Release wheel built and installed non-editably
- `scripts/run_installed_pytest.sh tests/fast/test_datasink.py` — 51 passed on the final rebased tree
- `scripts/run_release_tests.sh` — 614 non-Ray tests and 12 real-Ray tests passed
- `VCPKG_INSTALLED_DIR=<pinned-vcpkg-root> VANE_NATIVE_BUILD_JOBS=32 scripts/run_native_tests.sh "[distributed]"` — 234 cases / 9074 assertions passed
- `scripts/run_fast_tests.sh`:
  - non-Ray shard — 6451 passed, 602 skipped, 127 deselected, 7 xfailed, 1 xpassed
  - shared-Ray shard — 110 passed, 6 skipped, 7072 deselected
  - test-owned Ray clusters — 7 passed; vLLM-only test skipped because vLLM is not installed

## Checklist

- [x] The change is focused and includes tests or a reason tests are unnecessary.
- [x] Public behavior and compatibility impact are documented.
- [x] New dependencies, copied code, model assets, and datasets have compatible
      licenses and are recorded where required.
- [x] No credentials, private endpoints, personal paths, generated data, model
      weights, or build artifacts are included.
- [x] Security implications of UDFs, serialization, remote code, network access,
      and untrusted input have been considered.
- [x] Native changes were compiled; Python-only, documentation, and workflow
      changes passed relevant checks.

### CODE REVIEW

**Comment by kaka11chen** [comment]
@codex review

File: vane/runners/ray/driver.py [hunk 1]
```diff
@@ -8434,6 +8669,51 @@ def stream_plan(
...
+            plan_id,
+            _time.time(),
+        )
+        result: dict[str, Any] | None = None
+        try:
+            resolved = progress.resolve(future)
+            if not isinstance(resolved, Mapping):
+                raise TypeError(f"Ray DataSink returned {type(resolved).__name__}, expected a mapping")
+            result = dict(resolved)
+        except BaseException as execution_error:
```
Comment (hubgeter) [inline vane/runners/ray/driver.py:9143]: This exception path only finalizes the progress renderer; it never cancels/settles `future` or closes `plan_id`. A Ray ObjectRef timeout or transport failure can therefore leave the DataSink plan, lifecycle registration, and query resources alive while `write_datasink` returns UNKNOWN. Please mirror `_teardown_failed_plan()` here (cancel/settle the remote call, then invoke `close_plan`) and add a regression test covering a failure after execution has started.

File: external/duckdb/src/include/duckdb/execution/distributed/plan/runner.hpp [hunk 2]
```diff
@@ -795,6 +817,27 @@ class PlanRunner : public std::enable_shared_from_this<PlanRunner> {
...
+			result.operation_id = data_sink_node->operation_id();
+			auto parsed = data_sink_node->finalize(collected_partitions);
+			if (parsed.is_ok()) {
+				result = std::move(parsed).value();
+			} else {
+				errors.push_back("selected worker results could not be parsed: " + string(parsed.error().what()));
+			}
+			result.outcome_aborted = false;
+			result.outcome_unknown = true;
+			result.outcome_error = StringUtil::Join(errors, "; ");
```
Comment (hubgeter) [inline external/duckdb/src/include/duckdb/execution/distributed/plan/runner.hpp]: `outcome_error` is assembled from unbounded `what()` strings from the primary error, abort barrier, and result parser. It is forwarded through pybind before Python applies its 4 KiB diagnostic truncation, so the Ray result can exceed the bounded outcome-payload contract. Please cap diagnostic strings at the native/pybind boundary and add an oversized-error regression test.

**Review by hubgeter — CHANGES_REQUESTED** [review]
Request changes.

The provider-neutral DataSink contract and the APPLIED/ABORTED/UNKNOWN result model are a useful foundation. I reviewed head `d4e0507d029717cf590f6dd239352cc55d54eeaf` against base `bd063236e172c3a50e7d0845a4e8843dffc35114`.

I found two issues that should be addressed before merge:

1. The Ray client exception path does not cancel/settle the remote DataSink call or close the plan after a timeout or transport failure. This can leave the plan lifecycle and query resources owned by the session while the caller already receives an UNKNOWN outcome.
2. Native `outcome_error` diagnostics are not bounded before crossing the Ray/pybind boundary, so oversized worker/provider errors can violate the bounded outcome-payload contract.

The exact-head diff checks, Ruff, and Python compilation passed locally. The required CI run [32611445623](https://github.com/AstroVela/vane/actions/runs/32611445623) is green, but it tested the pull-request merge ref built from the old base. Current `main` is `ef5ff3b6db9404df21db90b7ca061a488714546a`; please update/rebase this branch and rerun the required native and real-Ray checks after fixing the findings.

File: vane/runners/ray/driver.py (hunk 1 above)
Comment (kaka11chen) [reply vane/runners/ray/driver.py:9143]: Fixed in 2ff92a451d. RayQueryDriverClient.run_datasink_plan now routes post-submission resolve failures through _teardown_failed_plan(), which cancels and settles the ObjectRef before invoking close_plan. The regression test asserts the initial failed wait, cancellation, bounded settle, and plan close order for a response timeout after submission.

File: external/duckdb/src/include/duckdb/execution/distributed/plan/runner.hpp (hunk 2 above)
Comment (kaka11chen) [reply external/duckdb/src/include/duckdb/execution/distributed/plan/runner.hpp]: Fixed in 2ff92a451d. DataSink outcome diagnostics are now UTF-8 sanitized and capped at 4096 bytes in PlanRunner, with a defensive cap on both pybind result paths. The native regression test uses an oversized error and verifies the byte limit plus preserved head/tail context. The full native [distributed] suite passed: 231 cases / 9052 assertions.

**Comment by kaka11chen** [comment]
@codex review

**Review by hubgeter — APPROVED** [review]

**Comment by kaka11chen** [comment]
Actor-only follow-up at `a1747d1ca6`:

- Replaced the task/actor mode split with the public `DataSinkWorker` actor contract; no compatibility aliases or task fallback remain.
- Each actor lazily owns one worker across batches and deterministically closes it during graceful local or Ray teardown; failed workers abort and cannot be reused.
- Kept backpressure and RefBundle transport in the shared distributed UDF layer. `worker_count` bounds sink concurrency, while workers continue to receive materialized `pyarrow.Table` batches.
- Rebuilt the Release wheel with 32-way parallelism. Targeted tests, 598 non-Ray release tests, and 12 real-Ray release tests passed.

@codex review

File: vane/execution/udf_ray_actor_pool.py [hunk 3]
```diff
@@ -200,17 +217,37 @@ def shutdown(self, *, kill: bool = False) -> None:
...
             try:
                 ray.kill(actor, no_restart=True)
             except BaseException as exc:
-                errors.append(f"actor-{actor_index}: {type(exc).__name__}: {exc}")
+                kill_errors.append(f"actor-{actor_index}: {type(exc).__name__}: {exc}")
                 remaining_actors.append(actor)
         self.actors = remaining_actors
-        if errors:
-            raise RuntimeError("failed to shut down UDF actor pool: " + "; ".join(errors))
+        _warn_actor_close_errors(close_errors)
```
[the remark this answers is not in the document]
Comment (kaka11chen) [reply vane/execution/udf_ray_actor_pool.py]: Fixed in fbe05270a5. `UDFActorPoolBase.shutdown()` now raises when graceful callable cleanup fails even after Ray successfully terminates every actor. The driver distinguishes that completed-cleanup diagnostic from retryable kill failures, releases the terminated pool, preserves phase-retirement diagnostics through later cleanup/session-close failures, and adds them to `data_sink_cleanup_warnings`. The real-Ray DataSink regression verifies an APPLIED result contains the planned worker-close warning.

File: vane/datasink/__init__.py [hunk 4]
```diff
@@ -0,0 +1,1043 @@
...
+def _relation_arrow_schema(relation: DuckDBPyRelation) -> pa.Schema:
+    import pyarrow as pa
+
+    from vane.datasource import _convert_duckdb_pytype
+
+    names = list(relation.columns)
+    types = list(relation.types)
+    if len(names) != len(types):
+        raise RuntimeError("relation returned inconsistent column names and types")
+    return pa.schema([pa.field(name, _convert_duckdb_pytype(dtype)) for name, dtype in zip(names, types, strict=True)])
```
[the remark this answers is not in the document]
Comment (kaka11chen) [reply vane/datasink/__init__.py]: Fixed in fbe05270a5. DataSink binding now obtains the relation schema from a native `_arrow_schema()` method that uses DuckDB Arrow conversion with the relation/result client properties, without executing the relation. Coverage verifies DECIMAL(10,2), UUID, TIMETZ, and ENUM map to the same Arrow types used for worker batches.

File: vane/runners/local/runner.py [hunk 5]
```diff
@@ -551,3 +551,147 @@ def execute_write() -> dict[str, Any]:
...
+                if finalization_error is not None:
+                    raise finalization_error
+        finally:
+            primary_error = sys.exc_info()[1]
+            cleanup_errors = _shutdown_local_write_resources(
+                backend,
+                fragment_executor,
+                conn,
+                udf_actor_pools,
+                kill_actor_pools=result is None or result.get("outcome_unknown") is True,
```
[the remark this answers is not in the document]
Comment (kaka11chen) [reply vane/runners/local/runner.py]: Fixed in fbe05270a5. Local teardown now uses graceful actor shutdown after both the backend and fragment executor have quiesced, including ordinary UNKNOWN write failures. It uses forced actor termination only when backend or fragment quiescence cannot be established and keeps the connection alive in that unsafe path. The integration regression verifies a failing worker runs `abort()` followed by `close()`.

**Comment by kaka11chen** [comment]
@codex review

**Comment by kaka11chen** [comment]
Follow-up `c01ddfdfbf` fixes the code-quality mypy failure by using distinct names for the `list[str]` error details and the rendered `str` cleanup diagnostics. This is a type-only/local-name change with no runtime behavior change.

Validation after two consecutive clean review rounds:

- CI-equivalent changed-file pre-commit checks, including mypy: passed
- Release incremental native install with `CMAKE_BUILD_PARALLEL_LEVEL=32`: passed
- affected non-real-Ray tests: 679 passed, 10 deselected
- affected real-Ray DataSink test: 1 passed, 46 deselected
- release suite: 607 passed, 16 deselected; real-Ray phase: 12 passed, 611 deselected
- complete fast suite: 6444 passed / 602 skipped / 127 deselected / 7 xfailed / 1 xpassed; shared-Ray: 110 passed / 6 skipped; owner-Ray: 7 passed

All review threads remain resolved.

**Comment by kaka11chen** [comment]
@codex review

File: external/duckdb/src/include/duckdb/execution/distributed/plan/runner.hpp [hunk 6]
```diff
@@ -930,12 +973,23 @@ class PlanRunner : public std::enable_shared_from_this<PlanRunner> {
...
 			if (deferred_collection_error) {
+				if (data_sink_node) {
+					return data_sink_unknown_after_worker_abort(*deferred_collection_error);
+				}
 				return fail_after_worker_abort(*deferred_collection_error);
 			}
 
+			if (data_sink_node) {
+				auto finalize_res = data_sink_node->finalize(collected_partitions);
+				if (finalize_res.is_err()) {
```
[the remark this answers is not in the document]
Comment (kaka11chen) [reply external/duckdb/src/include/duckdb/execution/distributed/plan/runner.hpp:1064]: Fixed in 7f798ea235. DataSink execution now uses a capacity-one result channel and the worker managers publish selected-attempt outputs incrementally. Each selected Ray handle is acknowledged and released immediately after publication, while loser attempts are never published. DataSinkResultCollector checks the exact one-million-result and 64 MiB logical wire budgets before retaining each result, preflights reported row counts before materialization, and closes the receiver before abort/quiesce on either returned or thrown collection errors so producers cannot remain blocked. Native coverage includes exact budget boundaries, early oversized-count rejection, non-retention, channel backpressure/close wakeup, and selected-versus-loser streaming behavior.

File: vane/runners/local/runner.py [hunk 7]
```diff
@@ -158,15 +157,20 @@ def _shutdown_local_write_resources(
...
     except BaseException as exc:
         errors.append(exc)
+        errors.extend(_shutdown_udf_actor_pools(actor_pools, kill=True))
         return errors
 
-    errors.extend(_shutdown_udf_actor_pools(actor_pools, kill=kill_actor_pools))
+    # Reaching actor cleanup proves that backend and fragment execution have
+    # quiesced. Let resident callables release provider state before their
+    # processes exit, including after an ordinary write failure.
+    errors.extend(_shutdown_udf_actor_pools(actor_pools, kill=False))
```
[the remark this answers is not in the document]
Comment (kaka11chen) [reply vane/runners/local/runner.py]: Fixed in 7f798ea235. _SingleSubprocessExecutor._close_locked() now reports control-channel timeout/disconnect and post-ACK process-exit timeout as cleanup errors after still killing the subprocess and releasing all resources. Local and local-fast DataSink writes therefore preserve an APPLIED result while adding the failed graceful close to data_sink_cleanup_warnings. The regression uses a blocking DataSinkWorker.close() and verifies the bounded timeout, escalation, cleanup, and returned warning.

File: vane/runners/ray/driver.py [hunk 8]
```diff
@@ -7145,6 +7298,180 @@ def discard_completed_cleanup(completed: asyncio.Task[CopyPlanOutcome]) -> None:
...
+                self._teardown_plan_resources,
+                plan_id,
+            )
+        except asyncio.CancelledError:
+            pass
+        except BaseException as error:
+            teardown_error = error
+        if teardown_error is not None:
+            add_cleanup_warning(public_result, "DataSink teardown", teardown_error)
+        return public_result
```
[the remark this answers is not in the document]
Comment (kaka11chen) [reply vane/runners/ray/driver.py:7820]: Fixed in 7f798ea235. Ray DataSink teardown now distinguishes a completed _QueryCleanupDiagnostic from an incomplete teardown. Retryable teardown failures schedule one background retry per plan-lifecycle identity with bounded exponential backoff; every attempt is identity-fenced to prevent an old retry from tearing down a reused plan ID, and runtime shutdown awaits outstanding cleanup tasks. The successful write is returned once with a cleanup warning and is never re-executed. Tests cover two failed teardowns followed by successful release and the same-plan-ID ABA case.

**Comment by kaka11chen** [comment]
Addressed the three latest review findings in 7f798ea235:

- Timed-out local DataSinkWorker.close() now surfaces as a cleanup warning after forced process cleanup.
- Incomplete Ray DataSink teardown is retried per lifecycle with identity fencing and no write re-execution.
- Selected DataSink results now stream through a capacity-one channel, release Ray handles incrementally, and enforce exact count/wire-size caps before retention.

All three review threads have been replied to and resolved.

Validation completed after two consecutive clean review rounds: changed-file format and diff checks; Release build with 32-way native parallelism; 935 affected tests; release suite (611 non-Ray + 12 Real-Ray); full native distributed suite (234 cases / 9074 assertions); and the full fast suite (6448 non-Ray + 110 shared-Ray + 7 isolated-Ray passes, with expected environment skips).

**Comment by kaka11chen** [comment]
@codex review

File: src/vane_py/python_udf_actor_resources.cpp [hunk 9]
```diff
@@ -198,7 +198,7 @@ class PythonUDFActorResourceState : public ClientContextState {
 			return;
 		}
 		PythonGILWrapper gil;
-		ShutdownResources();
+		ShutdownResources(error && error->HasError());
```
[the remark this answers is not in the document]
Comment (kaka11chen) [reply src/vane_py/python_udf_actor_resources.cpp]: Fixed in d14f9fca86. ClientContext drains Executor::CancelTasks before QueryEnd callbacks, so QueryEnd now always calls ShutdownResources(false); kill=true remains only for setup rollback and context destruction. Added test_local_fast_datasink_worker_failure_closes_after_abort, which asserts pool shutdown received false and the worker observed abort followed by close.

[the remark this answers is not in the document]
Comment (kaka11chen) [reply src/vane_py/python_udf_actor_resources.cpp]: Fixed in d14f9fca86. QueryEnd now captures bounded Python shutdown diagnostics and transfers them through relation-local state into DataSink summaries without replacing the primary result or error. Capture is armed only after the current plan is prepared, preventing an older streaming query from leaking warnings. Added successful-close-failure and graceful-timeout regressions; both remain APPLIED with cleanup warnings.

**Comment by kaka11chen** [comment]
@codex review

**Comment by kaka11chen** [comment]
Addressed the latest review findings in d14f9fca86:

- failed local-fast DataSink queries now quiesce and gracefully close actor pools, so workers observe abort followed by close; forced shutdown remains limited to setup rollback, destruction, and graceful timeout escalation
- local-fast actor shutdown exceptions are captured as bounded, relation-local cleanup warnings and preserved on APPLIED, ABORTED, and UNKNOWN results without replacing the primary result/error
- warning capture is armed only after the current relation is prepared, preventing cleanup from an older streaming query from leaking into the new relation
- added regressions for failed-write kill mode, close exceptions, and graceful-close timeout diagnostics

The branch was rebased onto upstream/main at 01f70c5e34 and is 0 commits behind. Final validation includes a 32-way Release build, 51/51 DataSink tests on the rebased tree, release gates, and the complete fast launcher. Both review threads were replied to and resolved.

File: vane/runners/ray/driver.py [hunk 10]
```diff
@@ -2310,6 +2353,9 @@ async def _shutdown_runtime(self) -> None:
                 plan_runner = self.plan_runner
             await self.stop_client_lease_maintenance()
             await self.stop_query_resource_maintenance()
+            cleanup_tasks = tuple(self._datasink_cleanup_tasks.values())
+            if cleanup_tasks:
+                await asyncio.gather(*cleanup_tasks, return_exceptions=True)
```
[the remark this answers is not in the document]
Comment (kaka11chen) [reply vane/runners/ray/driver.py]: Fixed in 4e543bf02f. Runtime shutdown now fences and cancels DataSink cleanup retries, waits only for the bounded shutdown timeout, removes their registry entries, and then runs plan_runner.shutdown on the independently drained session executor. test_query_driver_shutdown_bounds_in_progress_datasink_cleanup covers a cleanup task that ignores cancellation.

File: vane/datasink/__init__.py [hunk 11]
```diff
@@ -0,0 +1,1047 @@
...
+        **options.map_batches_kwargs(runner_type),
+    )
+    terminal = mapped._mark_datasink(context.operation_id)
+
+    results: tuple[WriteResult, ...] = ()
+    warnings: tuple[str, ...] = ()
+    try:
+        if runner_type == "local-fast":
+            try:
+                table = terminal.to_arrow_table()
```
[the remark this answers is not in the document]
Comment (kaka11chen) [reply vane/datasink/__init__.py:1169]: Fixed in 4e543bf02f. Local execution now installs a native PhysicalDataSink that validates result count, per-field size, and the cumulative 64 MiB wire budget incrementally as chunks stream, before result-table materialization. DataSinkResultCollector also rejects reported partition and bundle sizes before Materialize. Native tests cover both streaming rejection and pre-materialization bundle rejection.

File: vane/datasink/__init__.py [hunk 12]
```diff
@@ -0,0 +1,1047 @@
...
+                raise _aborted_error(
+                    context,
+                    results,
+                    "keyed DataSink input validation rejected the operation before workers opened",
+                    warnings,
+                )
+            if WriteState.ABORTED in states:
+                raise _unknown_error(context, results, "DataSink results mixed applied and aborted states", warnings)
+            return _summary(context, WriteOutcome.APPLIED, results, warnings)
+        native_result = get_or_create_runner().run_datasink(terminal)
```
[the remark this answers is not in the document]
Comment (kaka11chen) [reply vane/datasink/__init__.py:1196]: Fixed in 4e543bf02f. write_datasink now validates the source relation before binding, validates the relation returned by prepare_input, and the terminal relation rechecks the transaction at execution. Tests cover source, late-bound terminal, and replacement prepared-relation explicit transactions.

**Comment by kaka11chen** [comment]
@codex review

**Review by hubgeter — APPROVED** [review]
